sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/SingletonPatternCheck.kt

```diff
@@ -0,0 +1,131 @@
+import org.jetbrains.kotlin.psi.psiUtil.visibilityModifier
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.AbstractCheck
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+import org.sonarsource.kotlin.visiting.KtTreeVisitor
+
+@Rule(key = "S6515")
+class SingletonPatternCheck : AbstractCheck() {
+
+    override fun visitKtFile(file: KtFile, kotlinFileContext: KotlinFileContext) {
+        val singletonClassCandidateExtractor = SingletonClassCandidateExtractor()
+        singletonClassCandidateExtractor.visitTree(file)
+        val singletonClassCandidates = singletonClassCandidateExtractor.singletonClassCandidates
+        if (singletonClassCandidates.isEmpty()) return
+
+        val singleConstructorCallExtractor =
+            SingleConstructorCallExtractor(singletonClassCandidates, kotlinFileContext.bindingContext)
+        singleConstructorCallExtractor.visitTree(file)
+        val singleConstructorCallByClass = singleConstructorCallExtractor.singleConstructorCallByClass
+        if (singleConstructorCallByClass.isEmpty()) return
+
+        singleConstructorCallByClass.values.stream().filter {
+            isInitializingCall(it) || isLazyInitializingCall(it, kotlinFileContext.bindingContext)
+        }.forEach {
+            kotlinFileContext.reportIssue(it, "Singleton pattern should use object declarations or expressions")
+        }
+    }
+}
+
+private class SingletonClassCandidateExtractor : KtTreeVisitor() {
+
+    val singletonClassCandidates = mutableSetOf<String>()
+
+    override fun visitClass(klass: KtClass) {
+        if (isSingletonClassCandidate(klass)) {
+            klass.fqName?.let {
+                singletonClassCandidates.add(it.toString())
+            }
+        }
+    }
+}
+
+private fun isSingletonClassCandidate(klass: KtClass): Boolean {
+    val constructors = klass.allConstructors
+    return klass.isPrivate() || (constructors.isNotEmpty() && constructors.all(KtModifierListOwner::isPrivate))
+}
+
+private fun KtModifierListOwner.isPrivate(): Boolean = visibilityModifier()?.text == "private"
+
+private class SingleConstructorCallExtractor(
+    private val singletonClassCandidates: MutableSet<String>,
+    private val bindingContext: BindingContext,
+) : KtTreeVisitor() {
+
+    val singleConstructorCallByClass: MutableMap<String, KtCallExpression> = mutableMapOf()
+
+    override fun visitCallExpression(expression: KtCallExpression) {
+        val fqName = getCalledConstructorClassFqNameOrNull(expression) ?: return
+        if (!singletonClassCandidates.contains(fqName)) return
+        if (singleConstructorCallByClass.put(fqName, expression) != null) {
+            singleConstructorCallByClass.remove(fqName)
+            singletonClassCandidates.remove(fqName)
+        }
+    }
+
+    private fun getCalledConstructorClassFqNameOrNull(expression: KtCallExpression): String? {
```

Comment:
Instead of using this custom logic, you can use a `FunMatcher` that matches the constructor you are looking for. `FunMatcher`s also try to run the logic in an optimized way, e.g. checking the number of arguments before resolving too much using semantics.

Reply:
Fixed.